[c10d][fr] Allow multiple writer registration with warnings

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150232

The life span of writer is actually the whole program which is sub-optimal but it is a practical compromise so that the registration of writer can happen outside PG creation.

So we decide to allow multiple writer registrations with warnings.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[c10d][fr] Allow multiple writer registration with warnings

[ghstack-poisoned]

d4l3k reviewed (approved):
LGTM

d4l3k commented:
Would be nice to add a unit test

kwen2501 commented on torch/csrc/distributed/c10d/FlightRecorder.cpp:
```diff
@@ -203,10 +203,12 @@ DebugInfoWriter& DebugInfoWriter::getWriter(int rank) {
-      "debugInfoWriter already registered");
+  if (hasWriterRegistered_.load()) {
+    TORCH_WARN_ONCE(
+      "DebugInfoWriter is registered more than once, but since we need it to stay "
+      "outside ProcessGroup, user needs to ensure that this registration is indeed needed. "
+      "And we will only use the last registered writer.");
```
How about:
"DebugInfoWriter has already been registered, ignoring"

fduwjj replied:
Well, the logic here is let the later registration overriding the old one. lol.

kwen2501 reviewed (approved):
lgtm, minor comment

fduwjj commented:
@d4l3k We do have a cpp unit test but it get disabled.. Let me also try to revive it in this PR.

fduwjj commented:
Looks like the test is still enabled. I will merge this PR.

fduwjj committed:
Update on "[c10d][fr] Allow multiple writer registration with warnings"


The life span of writer is actually the whole program which is sub-optimal but it is a practical compromise so that the registration of writer can happen outside PG creation.

So we decide to allow multiple writer registrations with warnings.

cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj commented:
Looks like failed tests mostly are related to protobuffer or is disabled test. Not related to this change.

fduwjj commented:
@pytorchbot merge -i
